### PR TITLE
MF-1257 - Access messages from readers endpoint with user access token

### LINKED ISSUE #1257: Reader endpoint that retrieves messages related to user
**FEATURE REQUEST**

1. Is there an open issue addressing this request? If it does, please add a "+1" reaction to the
   existing issue, otherwise proceed to step 2.

2. Describe the feature you are requesting, as well as the possible use case(s) for it.
Currently there is endpoint that retrieves messages from channels that are connected to thing passed in request.
There should be endpoint where you can retrieve messages for channels that belong to user filtered by some criteria.

3. Indicate the importance of this feature to you (must-have, should-have, nice-to-have).

Must have

### PR DESCRIPTION
**Pull request by mteodor**
### What does this do?
Enables user to read messages from channels he owns

### Which issue(s) does this PR fix/relate to?
Resolves #1257.

### List any changes that modify/break current functionality
None
### Did you document any new/modified functionality?
No

### CODE REVIEW

File: cmd/cassandra-reader/main.go
Comment (manuio) [inline cmd/cassandra-reader/main.go]: Here and for all readers I'd move this log after to succeed the connection. something like `Established gRPC connection with to auth service`

**Review by manuio — CHANGES_REQUESTED** [review]

File: readers/api/transport.go
Comment (manuio) [inline readers/api/transport.go]: Dead code

File: readers/api/transport.go
Comment (manuio) [inline readers/api/transport.go]: Remove this log

**Review by manuio — CHANGES_REQUESTED** [review]

File: readers/auth.go
Comment (dborovcanin) [inline readers/auth.go]: I would say this abstraction is unnecessary. We can simply have two clients (`mainflux.ThingsServiceClient` and `mainflux.AuthServiceClient`) and chose between them depending on the context.

File: readers/api/transport.go
Comment (dborovcanin) [inline readers/api/transport.go]: I'd prefer if we put authorization type somewhere ( `Authorization` header prefix or in a separate header or something like that) so that we know which method to use, rather than implicitly fallback to user if thing authorization fails.

**Review by dborovcanin — CHANGES_REQUESTED** [review]

File: readers/api/transport.go
Comment (mteodor) [reply readers/api/transport.go]: That sounds ok to me, I propose the following 
`Authorization: <type> <value>`
So for thing key - `Authorization: Thing <value>`
And for user token - `Authorization: Bearer <value>`
What do you say?
Seems like the right way - https://stackoverflow.com/a/43164958/2873439
@drasko ?

Comment (dborovcanin) [reply readers/api/transport.go]: Makes sense. The problem is that the prefix will break compatibility with the current authorization header. We can support both (`Things` prefix and no prefix for Thing authorization, `Bearer` for user) to give some space for migration and make prefix required in the next release. Despite the most popular schemas being `Basic` and `Bearer`, I don't see anything wrong with the custom `Thing` schema for thing authorization.

File: readers/api/transport.go
Comment (dborovcanin) [inline readers/api/transport.go]: Use `strings.TrimPrefix`.

File: readers/api/transport.go
Comment (dborovcanin) [inline readers/api/transport.go]: Consider using an empty `switch-case`, like:
```go
	switch {
	case strings.HasPrefix(token, thingToken):
		.....
	case strings.HasPrefix(token, userToken):
		.....
	default:
		return errors.Wrap(errUnauthorizedAccess, errWrongToken)
	}
```
Looks like a nice fit.

File: readers/messages.go
Comment (dborovcanin) [inline readers/messages.go]:
```go
import "errors"
```

**Review by dborovcanin — CHANGES_REQUESTED** [review]

**Review by dborovcanin — DISMISSED** [review]

File: api/readers.yml
Comment (manuio) [inline api/readers.yml]: Are you sure about this section? It shoudl be
```
      description: | 
          Thing or User access token: 
          * For thing access use "Authorization: Thing <thing_key>"
          * For user access use "Authorization: Bearer <user_token>"
```

File: readers/api/transport.go
Comment (drasko) [inline readers/api/transport.go]: I would call it `ac` not `uc`

**Review by drasko — CHANGES_REQUESTED** [review]

File: readers/api/transport.go
Comment (drasko) [inline readers/api/transport.go:56]: Aren't users authenticated via Users service? Why using `AuthServiceClient` then, which b the name should connect to Auth service (which can't identify the user, just read the policies)?

**Review by drasko — CHANGES_REQUESTED** [review]

File: readers/api/transport.go
Comment (mteodor) [reply readers/api/transport.go:56]: we are using `AuthServiceClient`  to check the token passed as `Authorization` header in HTTP request, token that was issued by `auth` when user was authenticated

**Comment by drasko** [comment]
@mteodor what would be the best use-case where user, with his account, needs to read data, and he cannot create one app to which he (or many other users) can give api keys?

I am asking this, because in my opinion this can be solved with one app (Thing) and API keys features that we already have.

File: cmd/cassandra-reader/main.go
Comment (dborovcanin) [inline cmd/cassandra-reader/main.go:83]: How do we set this, since there is no env var added? Also, when you load config, this field should be populated. Please check this for all the other readers.

**Review by dborovcanin — CHANGES_REQUESTED** [review]

**Comment by mteodor** [comment]
>

**Comment by mteodor** [comment]
> I am asking this, because in my opinion this can be solved with one app (Thing) and API keys features that we already have.

@drasko, API keys will not work without this change

File: readers/api/endpoint_test.go
Comment (dborovcanin) [inline readers/api/endpoint_test.go]: Fix typo: `format`.

File: readers/api/transport.go
Comment (dborovcanin) [inline readers/api/transport.go]: The problem with this approach is that both Thing and User are `Bearers`, so the `Bearer` prefix makes sense in both cases.

Comment (mteodor) [reply readers/api/transport.go]: so what do you propose?
I'll remind you on this conversation https://github.com/mainflux/mainflux/pull/1470#discussion_r730734378

**Comment by drasko** [comment]
@dusanb94 please let's move this one forward

**Review by dborovcanin — CHANGES_REQUESTED** [review]
Please don't forget to update README files with the new env variable.

File: readers/influxdb/README.md
Comment (manuio) [inline readers/influxdb/README.md]: Do we need `MF_THINGS_AUTH_GRPC_URL` and `MF_THINGS_AUTH_GRPC_TIMEOUT` anymore? I'd even replace it in other services.

Comment (mteodor) [reply readers/influxdb/README.md]: yes we do

**Review by dborovcanin — DISMISSED** [review]

File: api/readers.yml
Comment (drasko) [inline api/readers.yml]: This is something that should be changed, probably.

File: readers/api/transport.go
Comment (drasko) [inline readers/api/transport.go]: Why there is an empty space after the word? Does not seem quite right... Anyway, we should probably not use custom header tagsm as authentication schemes are limited to [these](https://developer.mozilla.org/en-US/docs/Web/HTTP/Authentication#authorization_and_proxy-authorization_headers).

Comment (drasko) [reply readers/api/transport.go]: On the other hand - there are [variants](https://www.loginradius.com/blog/async/everything-you-want-to-know-about-authorization-headers/), and AWS is using [this](https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-auth-using-authorization-header.html) for example.

**Review by drasko — CHANGES_REQUESTED** [review]

File: readers/api/transport.go
Comment (mteodor) [reply readers/api/transport.go]: @dusanb94 and I already discussed this and I think we agreed with custom header authentication as I saw that it is widely being used ( and looks like AWS example testifies that ).
My initial idea was just to use `Authorization: <token_value|thing_key>` or  `Authorization:Bearer <token_value|thing_key>`, if one fails try other. If discard that solution I don't know any other way than to use different prefix to specify different authentication.

File: readers/api/transport.go
Comment (dborovcanin) [inline readers/api/transport.go]: This one looks like it's not used.

Comment (dborovcanin) [reply readers/api/transport.go]: It would be good to align with this: https://github.com/mainflux/mainflux/pull/1538#discussion_r784783227 (these errors will be a part of `pkg/errors/types.go` once #1538 is merged).

**Review by dborovcanin — CHANGES_REQUESTED** [review]

File: readers/api/transport.go
Comment (manuio) [inline readers/api/transport.go]: You can use `errors.ErrAuthentication` and remove this one

File: readers/api/transport.go
Comment (manuio) [inline readers/api/transport.go]: You are ignoring the custom errors returned by this method and they are probably already wrapped with `errors.Authorization`. Can you check it please? You can probably return `err` here.

File: readers/api/transport.go
Comment (manuio) [inline readers/api/transport.go]: You can inline here

File: readers/api/transport.go
Comment (manuio) [inline readers/api/transport.go]: You are always returning `errCannotAuthorizeUser`. Do we need to check if `e.Code() == codes.PermissionDenied`? Also, can we use `errors.ErrAuthorization` instead of the custom one?

**Review by manuio — CHANGES_REQUESTED** [review]

File: readers/api/transport.go
Comment (manuio) [inline readers/api/transport.go]: Consider moving this to validation in resquests.go

File: readers/api/transport.go
Comment (manuio) [inline readers/api/transport.go:212]: This switch case should be probaly moved  after the validation in endpoint layer

File: readers/api/requests.go
Comment (manuio) [inline readers/api/requests.go:24]: Looks like `chanID` check is missing

**Review by manuio — CHANGES_REQUESTED** [review]

File: readers/api/transport.go
Comment (drasko) [inline readers/api/transport.go:224]: Move this line of initialization below, in `if` statement.

File: readers/api/transport.go
Comment (drasko) [inline readers/api/transport.go:217]: Inline with the line below

**Review by drasko — CHANGES_REQUESTED** [review]

File: readers/api/transport.go
Comment (manuio) [inline readers/api/transport.go]: You can do:
```
case errors.Contains(err, errors.ErrInvalidQueryParams),
	errors.Contains(err, errors.ErrMalformedEntity):
		w.WriteHeader(http.StatusBadRequest)
```

File: bootstrap/service_test.go
Comment (dborovcanin) [inline bootstrap/service_test.go]: What's this?

File: readers/api/transport.go
Comment (dborovcanin) [reply readers/api/transport.go:217]: Actually, since `user` is used later, there is no need to inline. This is just fine:
```go

user, err = usersAuth.Identify(ctx, &mainflux.Token{Value: token})
if err != nil {
...}
use user here
....
```

File: readers/api/endpoint_test.go
Comment (dborovcanin) [inline readers/api/endpoint_test.go]: It's a good idea to introduce the type of client (thing or user) in the case description.

**Review by dborovcanin — CHANGES_REQUESTED** [review]

File: readers/api/transport.go
Comment (dborovcanin) [reply readers/api/transport.go:224]: For code readability, I would maybe keep it this way. I leave it to @mteodor to decide.

File: readers/api/transport.go
Comment (dborovcanin) [reply readers/api/transport.go]: @mteodor Please address this remark.

File: readers/api/transport.go
Comment (dborovcanin) [reply readers/api/transport.go]: @mteodor Please address this remark.

**Review by dborovcanin — CHANGES_REQUESTED** [review]

File: readers/api/transport.go
Comment (mteodor) [reply readers/api/transport.go:224]: I would leave it as it is

File: readers/api/transport.go
Comment (dborovcanin) [inline readers/api/transport.go:48]: These errors should be handled in the `encodeError` to return `StatuForbidden`, not `InternalServerError`(by default).

**Review by dborovcanin — CHANGES_REQUESTED** [review]

File: readers/api/transport.go
Comment (mteodor) [reply readers/api/transport.go:48]: these errors wrapped intot `ErrAuthorization` which results in `StatusUnauthorized`
```
if err := authorize(ctx, req, tc, ac); err != nil {
			return nil, errors.Wrap(errors.ErrAuthorization, err)
		}
```
```
	case errors.Contains(err, errors.ErrAuthentication):
		w.WriteHeader(http.StatusUnauthorized)
```

Comment (dborovcanin) [reply readers/api/transport.go:48]: Missed that one, thanks. 👍🏼

**Review by dborovcanin — APPROVED** [review]

**Review by manuio — APPROVED** [review]
LGTM

**Review by manuio — DISMISSED** [review]

**Review by dborovcanin — DISMISSED** [review]

File: readers/api/transport.go
Comment (drasko) [reply readers/api/transport.go]: This is OK, but it should be `"Thing"`, not `"Thing "`.

File: readers/api/transport.go
Comment (drasko) [inline readers/api/transport.go]: Same here - `"Bearer"`, not `"Bearer "`.

**Review by drasko — CHANGES_REQUESTED** [review]

**Review by dborovcanin — APPROVED** [review]

**Review by manuio — APPROVED** [review]
LGTM

**Review by drasko — APPROVED** [review]
LGTM

**Comment by drasko** [comment]
Thanks @mteodor - this was a big one! Great contribution.

**Pull request merged**
